### DISCUSSION TITLE
Memory Pools

### DISCUSSION THREAD

**Discussion by kris-rowe** [libocca/occa#644] [Experimental]
Have you used the new `memoryPool` feature? 
Provide feedback or share your use cases here!

See [Example 17](https://github.com/libocca/occa/tree/main/examples/cpp/17_memory_pool) for more details.

**Comment by nnnunnn** [comment]
I tried a snapshot of occa + libParanumal::solver::cns (from 24 Nov 2022) with memory pools. MS Visual Studio 2022 failed to build launcher kernels for mode::cuda, so I reverted to snapshots (occa + libP) from August 2022. Currently tuning lightweight iso-surfing for cns. I'll explore source of error once iso-surfing is ready.

**Reply by kris-rowe** [reply]
The recent OCCA releases haven't been tested thoroughly on Windows, so any bug reports or patches would be helpful. 
(Working Windows into the CI pipeline is on the todo list for this year.)

**Reply by nnnunnn** [reply]
**Progress**: I updated my local Visual Studio 2022 to v17.4, cloned a copy of current occa, and tried a few tests with launcher kernels and memoryPools. So far so good. Next I'll try solvers/bns + new isoSurfing on current libP. This should give feature::memoryPool a good workout.

**Comment by stgeke** [comment]
How is memory fragmentation handled?

**Reply by kris-rowe** [reply]
@noelchalmers can better answer, however, currently there is no automatic defragmentation.

There are a few issues&mdash;e.g., related to memory slices&mdash;which need to be resolved before this can be added.

**Reply by noelchalmers** [reply]
Oh, I was going to say that we do have defrag. The pool is defraged on a resize, either automatic or manual. The Example 17 code shows one case of this happening
